discord_bot/discord_bot.py

```diff
@@ -1,60 +1,88 @@
+import asyncio
 import json
 import requests
-import asyncio
 import websockets
 
+from .opcodes import Opcodes
+
+
 CONFIG_FILE_PATH = 'configs/config.json'
 DISCORD_GATEWAY_ENDPOINT = 'https://discordapp.com/api/v6/gateway'
 
 
 class DiscordBot:
-
+    """
+    Establishes a connection to the discord gateway and handles varied messages
+    """
     def __init__(self):
-
         # instance variables
         self.config = self.load_config()
         self.gateway_ws_url = self.get_gateway()
+        self.heartbeat_interval = None
+        self.last_seq = None
+        self.event_loop = asyncio.get_event_loop()
+        self.websocket = None
 
-        # actions below
-        asyncio.get_event_loop().run_until_complete(self.gateway_handler())
+        # actions
+        self.event_loop.run_until_complete(self.gateway_handler())
 
     @staticmethod
     def load_config():
+        """
+        Loads the configurations
+        :return: dict - the configurations
+        """
         with open(CONFIG_FILE_PATH) as f:
             return json.load(f)
 
     @staticmethod
     def get_gateway():
+        """
+        Caches a gateway value and retrieves a new URL
+        :return: gateway URL
+        """
         r = requests.get(DISCORD_GATEWAY_ENDPOINT)
         return r.json()['url']
 
-    async def handshake(self, websocket):
-        identity_obj = {
-            "token": self.config["bot_token"],
-            "properties": {
-                "$os": "windows",
-                "$browser": "disco",
-                "$device": "disco"
-            },
-            "compress": False,
-            "large_threshold": 250,
-            "shard": [1, 10]
-        }
-        await websocket.send(json.dumps({"op": 2, "d": identity_obj}))
+    async def send_json(self, payload):
+        await self.websocket.send(json.dumps(payload))
 
-    async def gateway_handler(self):
+    async def handshake(self, message):
+        """
+        When a websocket connection is opened, Hello payload is received. Then, an Identify payload is sent as
+        part of the handshake to authorize this client.
 
+        :param message: dict - hello payload
+        """
+        self.heartbeat_interval = message["d"]["heartbeat_interval"]
+        await self.heartbeat()
```

Comment:
Here's the bug; `await` waits for another coroutine to complete before continuing.

After doing some more research, I found some important things I didn't realize before:

`await coroutine()` waits until that coroutine completes before continuing
`self.loop.create_task(coroutine())` creates a task and queues it up
`asyncio.ensure_future(coroutine())` queues up a coroutine directly

So we should re-evaluate all of our `await` and `create_task`s and convert them to `asyncio.ensure_future`s!